pointer-masking-proposal.adoc

```diff
@@ -295,76 +298,133 @@ MASK
  +
 
 [%header, cols=1*]
 ,===
 spmbase[63:0]
 BASE
 ,===
-Figure 8a: Memory Tagging Extension register (_**spmbase**_) for RV64 (S-MODE) +
+Figure 8c: Memory Tagging Extension register (_**spmbase**_) for RV64 (S-MODE) +
  +
 
 [%header, cols=1*]
 ,===
 upmbase[63:0]
 BASE
 ,===
-Figure 8b: Memory Tagging Extension register (_**upmbase**_) for RV64 (U-MODE) +
+Figure 8d: Memory Tagging Extension register (_**upmbase**_) for RV64 (U-MODE) +
+ +
+
+[%header, cols=1*]
+,===
+mpmbase[127:0]
+BASE
+,===
+Figure 9a: Memory Tagging Extension register (_**mpmbase**_) for RV128 (M-MODE) +
+ +
+
+[%header, cols=1*]
+,===
+hpmbase[127:0]
+BASE
+,===
+Figure 9b: Memory Tagging Extension register (_**hpmbase**_) for RV128 (H-MODE) +
  +
 
 [%header, cols=1*]
 ,===
 spmbase[127:0]
 BASE
 ,===
-Figure 9a: Memory Tagging Extension register (_**spmbase**_) for RV128 (S-MODE) +
+Figure 9c: Memory Tagging Extension register (_**spmbase**_) for RV128 (S-MODE) +
  +
 
 [%header, cols=1*]
 ,===
 upmbase[127:0]
 BASE
 ,===
-Figure 9b: Memory Tagging Extension register (_**upmbase**_) for RV128 (U-MODE) +
+Figure 9d: Memory Tagging Extension register (_**upmbase**_) for RV128 (U-MODE) +
  +
 
 [[explanation]]
 ## Explanation
 
-_**MMTE**_ register fully two-fold function:
+_**MMTE**_ register fulfills two-fold function:
 
-1. Can be programmable by more privileged mode
+1. Can be programmable by more privileged mode (unless PM.Current bit is enabled)
 2. Performs status register function for the current privilege mode
 
-_PM_ bits from **_MMTE_** register are accessible in all modes (U/S/HS/M) and can be read to query the current status of PM feature (if enabled). However, only M-mode privileged code can set the value for _PM_ bits.
+_PM_ bits from **_MMTE_** register are accessible in all modes (U/S/HS/M) and can be read to query the current status of PM feature (if enabled). By default, only higher privileged code can set the value for _PM_ bits. However, higher privileged code can enable _PM.Current_ bit for lower privileged code. In such scenario, current privilege code has a possibility to self-manage its own configuration of _PM_ bits.
+If _PM.Enable_ bit is changing the status, corresponding _xTE_ bit is automatically updated.
+
+_xTE_ bits fulfills two-fold functions:
+
+1. Can be used to quickly disable _PM_ for the corresponding mode without changing current configuration.
+2. Can be used to modify which currently set-up _**xPMMASK**_ / _**xPMBASE**_ and _PM_ bits are enforced.
 
-_UTE_ bit can be read by any mode, but only more privileged modes (S, HS and M) can change the value of it.
+By default, the current CPU mode is using _**xPMMASK**_, _**xPMBASE**_ and _PM_ bits corresponding to it. When CPU is switching the mode, all _xTE_ bits are cleared except the one corresponding to the new mode. _xTE_ bit corresponding to that mode reflects the current configuration of the _PM.Enable_ bit.
```

Comment:
How S-mode could use UPMMASK/UPMBASE if according to the text on the one hand xTE bit reflects the current configuration of PM.Enable, but on the other - xTE is used to select xPMMASK/xPMBASE to calculate the address?

Reply:
S-mode could use UPMMASK/UPMBASE by clearing all xTE bits and setting UTE. However, enforcement of UPMBASE/UPMMASK will happen if PM.Enable for that mode is also enabled (S-mode can enable it by itself).